src/lib/foreign-field.ts

```diff
@@ -0,0 +1,496 @@
+      let z = Provable.witness(ForeignField, () => this.toConstant().inv());
+
+      // in unsafe mode, `witness` didn't constrain z to be a valid field element
+      if (unsafe) z.assertValidElement();
+
+      // check that x * z === 1
+      // TODO: range checks added by `mul` on `one` are unnecessary, since we already assert that `one` equals 1
+      let one = Snarky.foreignField.mul(this.value, z.value, pMl);
+      new ForeignField(one).assertEquals(new ForeignField(1));
+
+      return z;
+    }
+
+    // convenience methods
+
+    /**
+     * Assert equality with a ForeignField-like value
+     *
+     * ```ts
+     * x.assertEquals(0, "x is zero");
+     * ```
+     */
+    assertEquals(y: ForeignField | bigint | number, message?: string) {
+      try {
+        if (this.isConstant() && isConstant(y)) {
+          let x = this.toBigInt();
+          let y0 = toFp(y);
+          if (x !== y0) {
+            throw Error(`ForeignField.assertEquals(): ${x} != ${y0}`);
+          }
+        }
+        return Provable.assertEqual(ForeignField, this, ForeignField.from(y));
+      } catch (err) {
+        throw withMessage(err, message);
+      }
+    }
+
+    /**
+     * Check equality with a ForeignField-like value
+     *
+     * ```ts
+     * let isXZero = x.equals(0);
+     * ```
+     */
+    equals(y: ForeignField | bigint | number) {
+      if (this.isConstant() && isConstant(y)) {
+        return new Bool(this.toBigInt() === toFp(y));
+      }
+      return Provable.equal(ForeignField, this, ForeignField.from(y));
+    }
+
+    // bit packing
+
+    /**
+     * Unpack a field element to its bits, as a `Bool[]` array.
+     *
+     * This method is provable!
+     */
+    toBits(length = sizeInBits) {
+      checkBitLength('ForeignField.toBits()', length, sizeInBits);
+      let [l0, l1, l2] = this.toFields();
+      let limbSize = Number(limbBits);
+      let xBits = l0.toBits(Math.min(length, limbSize));
+      length -= limbSize;
+      if (length <= 0) return xBits;
+      let yBits = l1.toBits(Math.min(length, limbSize));
+      length -= limbSize;
+      if (length <= 0) return [...xBits, ...yBits];
+      let zBits = l2.toBits(Math.min(length, limbSize));
+      return [...xBits, ...yBits, ...zBits];
+    }
+
+    /**
+     * Create a field element from its bits, as a `Bool[]` array.
+     *
+     * This method is provable!
+     */
+    static fromBits(bits: Bool[]) {
+      let length = bits.length;
+      checkBitLength('ForeignField.fromBits()', length, sizeInBits);
+      let limbSize = Number(limbBits);
+      let l0 = Field.fromBits(bits.slice(0 * limbSize, 1 * limbSize));
+      let l1 = Field.fromBits(bits.slice(1 * limbSize, 2 * limbSize));
+      let l2 = Field.fromBits(bits.slice(2 * limbSize, 3 * limbSize));
+      let x = ForeignField.fromFields([l0, l1, l2]);
+      // TODO: can this be made more efficient? since the limbs are already range-checked
+      if (length === sizeInBits) x.assertValidElement();
```

Comment:
In particular, it could be the case that the output is 256 bits for example, but the actual value is greater than the foreign modulus (of secp256k1 for example)